text/0000-const-trait-impls.md

```diff
@@ -0,0 +1,902 @@
+This is both handled this way by the parser, and semantically what is meant here. The constness of the trait bound is affected,
+the trait bound itself exists either way.
+
+## Why do traits need to be marked as "const implementable"?
+
+### Default method bodies
+
+Adding a new method with a default body would become a breaking change unless that method/default body
+would somehow be marked as `const`, too. So by marking the trait, you're opting into the requirement that all default bodies are const checked,
+and thus neither `impl const Trait for Type` items nor `impl Trait for Type` items will be affected if you add a new method with a default body.
+This scheme avoids adding a new kind of breaking change to the Rust language,
+and instead allows everyone managing a public trait in their crate to continue relying on the
+previous rule "adding a new method is not a breaking change if it has a default body".
+
+### `~const Destruct` super trait
+
+Traits that have `self` (by ownership) methods, will almost always drop the `self` in these methods' bodies unless they are simple wrappers that just forward to the generic parameters' bounds.
+
+The following never drops `T`, because it's the job of `<T as Add>` to handle dropping the values.
+
+```rust
+struct NewType<T>(T);
+
+impl<T: ~const Add<Output = T>> const Add for NewType<T> {
+    type Output = Self,
+    fn add(self, other: Self) -> Self::Output {
+        NewType(self.0 + other.0)
+    }
+}
+```
+
+But if any code path could drop a value...
+
+```rust
+struct NewType<T>(T, bool);
+
+struct Error;
+
+impl<T: ~const Add<Output = T>> const Add for NewType<T> {
+    type Output = Result<Self, Error>;
+    fn add(self, other: Self) -> Self::Output {
+        if self.1 {
+            Ok(NewType(self.0 + other.0, other.1))
+        } else {
+            // Drops both `self.0` and `self.1`
+            Err(Error)
+        }
+    }
+}
+```
+
+... then we need to add a `~const Destruct` bound to `T`, to ensure
+`NewType<T>` can be dropped.
+
+This bound in turn will be infectious to all generic users of `NewType` like
+
+```rust
+const fn add<T: ~const Add>(
+    a: NewType<T>,
+    b: NewType<T>,
+) -> Result<NewType<T::Output>, Error> {
+    a + b
+}
+```
+
+which now need a `T: ~const Destruct` bound, too.
+In practice we have noticed that a large portion of APIs will have a `~const Destruct` bound.
+This bound has little value as an explicit bound that appears almost everywhere.
+Especially since it is a fairly straight forward assumption that a type that has const trait impls will also have a `const Drop` impl or only contain `const Destruct` types.
+
+Thus we give all `const trait`s a `~const Destruct` super trait to ensure users don't need to add `~const Destruct` bounds everywhere.
+We may offer an opt out of this behaviour in the future, if there are convincing real world use cases.
```

Comment:
Interesting. This can be seen as an extension of the fact that all vtables have a `drop` slot, so in a sense all traits already have the `Destruct` supertrait.